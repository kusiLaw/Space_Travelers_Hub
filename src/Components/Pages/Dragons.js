import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
// eslint-disable-next-line import/named
import { getDragons, bookDragon, cancelDragon } from '../../Redux/Dragons/Dragons';
import Dragon from '../Dragon';

export default function Dragons() {
  const dispatch = useDispatch();
  const dragons = useSelector((state) => state.dragons);

  useEffect(() => {
    if (dragons.length === 0) {
      dispatch(getDragons());
    }
  }, []);

  const handleBooking = (e) => {
    e.preventDefault();
    dispatch(bookDragon(e.target.id));
  };

  const handleCancelling = (e) => {
    e.preventDefault();
    dispatch(cancelDragon(e.target.id));
  };

  return (
    <section className="Dragons-container container-fluid d-md-block d-lg-flex gap-lg-3 mt-lg-5 mb-lg-5">
      {dragons.map((dragon) => (
        <Dragon
          key={dragon.id}
          dragon={dragon}
          id={dragon.id}
          handleBooking={handleBooking}
          handleCancelling={handleCancelling}
        />
      ))}
    </section>
  );
}
