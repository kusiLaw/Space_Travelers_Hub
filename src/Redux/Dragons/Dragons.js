import { createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const GET_DRAGON = 'GET_DRAGON';
const BOOK_DRAGON = 'BOOK_DRAGON';
const CANCEL_DRAGON = 'CANCEL_DRAGON';
const initialState = [];
const baseURL = 'https://api.spacexdata.com/v3/dragons';

const reducerDragons = (state = initialState, action) => {
  switch (action.type) {
    case `${GET_DRAGON}/fulfilled`:
      return [...state, ...action.payload];
    case `${BOOK_DRAGON}`:
      return state.map((dragon) => {
        if (dragon.id !== action.payload) return dragon;
        return { ...dragon, reserved: true };
      });
    case `${CANCEL_DRAGON}`:
      return state.map((dragon) => {
        if (dragon.id !== action.payload) return dragon;
        return { ...dragon, reserved: false };
      });
    default: return state;
  }
};

const getDragons = createAsyncThunk(GET_DRAGON,
  async () => {
    const response = await axios.get(baseURL);
    const data = await response.data.map((dragon) => ({
      id: dragon.id,
      designation: dragon.name,
      type: dragon.type,
      image: dragon.flickr_images[0],
      description: dragon.description,
      reserved: false,
    }));
    return data;
  });

const bookDragon = (id) => ({
  type: BOOK_DRAGON,
  payload: id,
});

const cancelDragon = (id) => ({
  type: CANCEL_DRAGON,
  payload: id,
});

export {
  reducerDragons,
  getDragons,
  bookDragon,
  cancelDragon,
};
