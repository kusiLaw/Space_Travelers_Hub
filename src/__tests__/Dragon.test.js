import { Provider } from 'react-redux';
import TestRenderer from 'react-test-renderer';
import store from '../Redux/configureStore';
import Dragons from '../Components/Pages/Dragons';

it('Testing Dragon component, should match snapshot:', () => {
  const tree = TestRenderer.create(
    <Provider store={store}>
      <Dragons />
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
