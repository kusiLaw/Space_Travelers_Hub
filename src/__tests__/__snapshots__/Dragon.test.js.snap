// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Dragon component, should match snapshot: 1`] = `
<section
  className="Dragons-container container-fluid d-md-block d-lg-flex gap-lg-3 mt-lg-5 mb-lg-5"
/>
`;
